explorer/src/db/mod.rs

```diff
@@ -0,0 +1,848 @@
+                discrimination,
+                prev_transactions: &transactions,
+                prev_blocks: &blocks,
+            },
+        );
+        let (stake_pool_data, stake_pool_blocks) =
+            apply_block_to_stake_pools(stake_pool_data, stake_pool_blocks, &explorer_block);
+
+        let stake_control = apply_block_to_stake_control(stake_control, &explorer_block);
+
+        let state_ref = multiverse
+            .insert(
+                chain_length,
+                block.parent_id(),
+                block_id,
+                State {
+                    transactions: apply_block_to_transactions(transactions, &explorer_block)?,
+                    blocks: apply_block_to_blocks(blocks, &explorer_block)?,
+                    addresses: apply_block_to_addresses(addresses, &explorer_block),
+                    epochs: apply_block_to_epochs(epochs, &explorer_block),
+                    chain_lengths: apply_block_to_chain_lengths(chain_lengths, &explorer_block)?,
+                    stake_pool_data,
+                    stake_pool_blocks,
+                    vote_plans: apply_block_to_vote_plans(
+                        vote_plans,
+                        &explorer_block,
+                        &stake_control,
+                    ),
+                    stake_control,
+                },
+            )
+            .await;
+
+        Ok(state_ref)
+    }
+
+    pub async fn get_block(&self, block_id: &HeaderHash) -> Option<Arc<ExplorerBlock>> {
+        for (_hash, state_ref) in self.multiverse.tips().await.iter() {
+            if let Some(b) = state_ref.state().blocks.lookup(block_id) {
+                return Some(Arc::clone(b));
+            }
+        }
+
+        None
+    }
+
+    pub async fn set_tip(&self, hash: HeaderHash) -> Result<(), BlockNotFound> {
+        // the tip changes which means now a block is confirmed (at least after
+        // the initial epoch_stability_depth blocks).
+
+        let state_ref = if let Some(state_ref) = self.multiverse.get_ref(&hash).await {
+            state_ref
+        } else {
+            return Err(BlockNotFound { hash });
+        };
+
+        let state = state_ref.state();
+        let block = Arc::clone(state.blocks.lookup(&hash).unwrap());
+
+        if let Some(confirmed_block_chain_length) = block
+            .chain_length()
+            .nth_ancestor(self.blockchain_config.epoch_stability_depth)
+        {
+            debug_assert!(
+                ChainLength::from(
+                    self.stable_store
+                        .confirmed_block_chain_length
+                        .load(Ordering::Acquire)
+                ) <= block.chain_length()
+            );
+
+            self.stable_store
+                .confirmed_block_chain_length
+                .store(confirmed_block_chain_length.into(), Ordering::Release);
+
+            self.multiverse
+                .gc(self.blockchain_config.epoch_stability_depth)
```

Comment:
Well, it's true that garbage collection is suboptimal with this. But the reasoning is that the data is still there, because each `State` is immutable, and we can't remove data from it, so the `State` at the tip has everything that you can query about the main branch.

The only thing you really lose with this `gc` is the possibility of using `branch(block_hash)` (which you can use to scope the state of the query), but if you do `tip { blocks_by_chain_length(0) }` you still get that block.

Reply:
Agree on the correctness part, but this has one additional problem which is that using `self.blockchain_config.epoch_stability_depth` is almost useless after a while, as it only prune blocks with chain length between 0 and that. I think we should use something like `current_stability_depth - epoch_stability_depth` like we do here https://github.com/input-output-hk/jormungandr/blob/f55155f71ebbb0c29eab51551c309c3367055eab/jormungandr/src/blockchain/storage.rs#L311

Reply:
> each State is immutable, and we can't remove data from it

This is also the reason adding a database to store only the non-volatile part is not that simple, btw. Or rather, garbage collecting the volatile part is the problematic part (especially for non-scalar indices like transactions by address)

Reply:
> Agree on the correctness part, but this has one additional problem which is that using self.blockchain_config.epoch_stability_depth is almost useless after a while, as it only prune blocks with chain length between 0 and that.

Are you sure?

`if let Some(gc_threshold_length) = longest_chain.nth_ancestor(cut_to_depth) {`

I see that line in that function, `longest_chain` should be increasing with each block, so the upper bound should be increasing too.

Reply:
https://github.com/input-output-hk/jormungandr/blob/f55155f71ebbb0c29eab51551c309c3367055eab/jormungandr/src/blockchain/chain.rs#L303

And we are also using it like that in jormungandr.

Reply:
you're right sorry, I don't know what I was looking at